packages/components/src/tabs/index.tsx

```diff
@@ -0,0 +1,157 @@
+/**
+ * External dependencies
+ */
+import * as Ariakit from '@ariakit/react';
+
+/**
+ * WordPress dependencies
+ */
+import { useInstanceId, usePrevious } from '@wordpress/compose';
+import { useEffect, useLayoutEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import type { TabsProps } from './types';
+import { TabsContext } from './context';
+import Tab from './tab';
+import TabList from './tablist';
+import TabPanel from './tabpanel';
+
+function Tabs( {
+	activeClass = 'is-active',
+	selectOnMove = true,
+	initialTabId,
+	orientation = 'horizontal',
+	onSelect,
+	children,
+	selectedTabId,
+}: TabsProps ) {
+	const instanceId = useInstanceId( Tabs, 'tabs' );
+	const store = Ariakit.useTabStore( {
+		selectOnMove,
+		orientation,
+		defaultSelectedId: initialTabId && `${ instanceId }-${ initialTabId }`,
+		setSelectedId: ( selectedId ) => {
+			const strippedDownId =
+				typeof selectedId === 'string'
+					? selectedId.replace( `${ instanceId }-`, '' )
+					: selectedId;
+			onSelect?.( strippedDownId );
+		},
+		selectedId: selectedTabId && `${ instanceId }-${ selectedTabId }`,
+	} );
+
+	const isControlled = selectedTabId !== undefined;
+
+	const { items, selectedId } = store.useState();
+	const { setSelectedId } = store;
+
+	const selectedTab = items.find( ( item ) => item.id === selectedId );
+	const firstEnabledTab = items.find( ( item ) => {
+		// Ariakit internally refers to disabled tabs as `dimmed`.
+		return ! item.dimmed;
+	} );
+	const initialTab = items.find(
+		( item ) => item.id === `${ instanceId }-${ initialTabId }`
+	);
+
+	// Handle selecting the initial tab.
+	useLayoutEffect( () => {
+		if ( isControlled ) {
+			return;
+		}
+
+		// Wait for the denoted initial tab to be declared before making a
+		// selection. This ensures that if a tab is declared lazily it can
+		// still receive initial selection, as well as ensuring no tab is
+		// selected if an invalid `initialTabId` is provided.
+		if ( initialTabId && ! initialTab ) {
+			return;
+		}
+
+		// If the currently selected tab is missing (i.e. removed from the DOM),
+		// fall back to the initial tab or the first enabled tab.
+		if ( ! items.find( ( item ) => item.id === selectedId ) ) {
+			if ( initialTab && ! initialTab.dimmed ) {
+				setSelectedId( initialTab?.id );
+			} else {
```

Comment:
Hmm, excellent point. This actually doesn't work as you'd expect it to.

<details>
<summary>Testing diff:</summary>

```diff
diff --git a/packages/components/src/tabs/index.tsx b/packages/components/src/tabs/index.tsx
index 3891f83753..ea052c6d09 100644
--- a/packages/components/src/tabs/index.tsx
+++ b/packages/components/src/tabs/index.tsx
@@ -75,6 +75,7 @@ function Tabs( {
 			if ( initialTab && ! initialTab.dimmed ) {
 				setSelectedId( initialTab?.id );
 			} else {
+				console.log( firstEnabledTab?.id );
 				setSelectedId( firstEnabledTab?.id );
 			}
 		}
diff --git a/packages/components/src/tabs/stories/index.story.tsx b/packages/components/src/tabs/stories/index.story.tsx
index b27c8bd42e..cce18d8209 100644
--- a/packages/components/src/tabs/stories/index.story.tsx
+++ b/packages/components/src/tabs/stories/index.story.tsx
@@ -364,10 +364,10 @@ const TabGetsRemovedTemplate: StoryFn< typeof Tabs > = ( props ) => {
 							Tab 1
 						</Tabs.Tab>
 					) }
-					<Tabs.Tab id={ 'tab2' } title={ 'Tab 2' }>
+					<Tabs.Tab id={ 'tab2' } title={ 'Tab 2' } disabled={ true }>
 						Tab 2
 					</Tabs.Tab>
-					<Tabs.Tab id={ 'tab3' } title={ 'Tab 3' }>
+					<Tabs.Tab id={ 'tab3' } title={ 'Tab 3' } disabled={ true }>
 						Tab 3
 					</Tabs.Tab>
 				</Tabs.TabList>

```

</details>

tl;dr: if there is no enabled tab, we end up calling `setSelectedId(undefined)`, and that leads to Ariakit not actually changing the selection, as far as I can tell. I _think_ this happens due to some [internal logic for handling `selectedId` being `undefined`](https://github.com/ariakit/ariakit/blob/901b2e3271c4fbbca81da4a87ef528fcad2f3e42/packages/ariakit-core/src/tab/tab-store.ts#L82C3-L100C5), but I could be reading that wrong.

We may be able to detect this case and explicitly call with `null`. I'll experiment a bit later tonight/in the am.

Reply:
Yeah, I see two alternatives:

- calling `setSelectedId( firstEnabledTab?.id )` only `if ( firstEnabledTab )` — which basically would result in the hook not performing any selection in case there isn't a `firstEnabledTab` 
- calling `setSelectedId( null )` in case there isn't a `firstEnabledTab`

Let's look into any potential difference between the two approaches, but the first one seems cleaner at least on paper.

Reply:
Calling `setSelectedId( null )` proved necessary in cases where there is no `firstEnabledTab`. Without doing so we wound up still sitting rendering the recently-removed tab, I think due to that internal fallback logic.

It was also necessary to make sure this call didn't happen too early, breaking other flows, because `firstEnableTab` starts out as `undefined` while the component figures out what it's working with during the initial render. I'm using a ref to track when the tab data initially populates so we can prevent this, and other effects, from triggering too early.

With ca702bab2d once the tabs populate for the first time, if the selected tab is removed and there are no other enabled tabs nothing will be selected.

Note: if you click the button in the story to replace the recently removed tab, it _will_ be selected automatically. This is not because it was the most recently selected tab though, it's just because in uncontrolled mode we always try to fall back to the first enabled tab. This auto-selection is expected.

Reply:
This all sounds reasonable — do we need to add unit tests for this new behavior (both controlled/uncontrolled as needed)?